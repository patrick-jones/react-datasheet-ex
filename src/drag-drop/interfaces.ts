import RDS from 'react-datasheet';
import {ConnectDragPreview, ConnectDragSource, ConnectDropTarget} from 'react-dnd';
import {ColumnHeader, HeaderRendererProps, RowRendererProps} from '../interfaces';

/**
 * Properties that must be provided to a drag and drop enabled row
 */
export interface DragDropRowProps<T extends RDS.Cell<T, V>, V = string> {
  /** The current row index */
  row: number;
  /** Callback for when a row is dropped on another */
  readonly onRowDropped: (source: number, target: number) => void;
}

/**
 * Properties that must be provided to a drag and drop enabled column header
 */
export interface DragDropHeaderProps {
  /** The current column index */
  col: number;
  /** Callback for one header is dropped on another */
  readonly onHeaderDropped: (source: number, target: number) => void;
}

/**
 * Drag source properties that are injected into a drag and drop provider
 */
export interface InjectedDragSourceProps {
  readonly connectDragSource: ConnectDragSource;
  readonly connectDragPreview: ConnectDragPreview;
}

/**
 * Drop target properties that are injected into a drag and drop provider
 */
export interface InjectedDropTargetProps {
  readonly connectDropTarget: ConnectDropTarget;
  readonly isDropOver: boolean;
}

/**
 * Combined drag source and drop target properties that are injected
 * into a drag and drop provider
 */
export interface InjectedDragDropProps extends InjectedDragSourceProps, InjectedDropTargetProps {}

/**
 * Extends `ReactDatasheet.RowRendererProps` to include
 * drag and drop properties
 */
export interface DragDropRowRendererProps<CellType, V = string>
  extends RowRendererProps<CellType, V>,
    DragDropRowProps<CellType, V> {}

/**
 * Extends `HeaderRendererProps` to include drag and drop properties.
 */
export interface DragDropHeaderRendererProps<T extends ColumnHeader = ColumnHeader>
  extends HeaderRendererProps<T>,
    DragDropHeaderProps {}
