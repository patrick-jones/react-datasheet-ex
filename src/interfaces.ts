import React, {CSSProperties} from 'react';
import RDS from 'react-datasheet';

export interface RowLocation {
  row: number;
}

export interface ColumnLocation {
  col: number;
}

export interface ColumnHeader {
  id: string;
  title: string;
  width?: number;
}

export interface HeaderRendererProps<T extends ColumnHeader = ColumnHeader> extends ColumnLocation {
  header: T;
}

export type HeaderRenderer<T extends ColumnHeader = ColumnHeader,
  R extends HeaderRendererProps<T> = HeaderRendererProps<T>> = React.ComponentType<R>;

export type HeaderContentRenderer<T extends ColumnHeader = ColumnHeader>
  = React.ComponentType<HeaderRendererProps<T>>;

export interface SheetRendererProps<
    T extends RDS.Cell<T, V>, V = string, H extends ColumnHeader = ColumnHeader
  >
  extends RDS.SheetRendererProps<T, V> {
  /** React style object that should be applied to the enclosing table */
  style?: CSSProperties | undefined;
  /**
   * An array of `ColumnHeader`, which are rendered into `th` elements
   * @default []
   */
  headers: H[];
  /**
   * Optional component or function to render the header cell contents.
   * The contents are rendered within a `th` element.
   * @default DefaultHeaderRenderer
   */
  headerRenderer: HeaderRenderer;
  /**
   * Optional component or function to render the header cell contents.
   * The contents are rendered within a `th` element.
   * @default DefaultHeaderContentRenderer
   */
  headerContentRenderer: HeaderContentRenderer;
}


export type ActionContentRenderer<T extends RowLocation = RowLocation>
  = React.ComponentType<T>;

export interface RowRendererProps<
    T extends RDS.Cell<T, V>, V = string, AP extends RowLocation = RowLocation
  > extends RDS.RowRendererProps<T, V> {

  actionContentRenderer?: ActionContentRenderer<AP>;
}
