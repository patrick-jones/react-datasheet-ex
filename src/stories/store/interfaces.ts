import RDS from 'react-datasheet';
import {ColumnHeader} from '../../interfaces';

export type ExampleCellValue = string | number | null;
export interface ExampleCellType extends RDS.Cell<ExampleCellType> {
  value: ExampleCellValue;
}

export type ExampleModel = {
  id: string;
  headers: ColumnHeader[];
  rows: ExampleRow[];
  selected: RDS.Selection | null;
};

export type ExampleRow = {
  id: string;
  selected?: boolean;
  data: ExampleCellType[];
};

export type Grid = ExampleCellType[][];
