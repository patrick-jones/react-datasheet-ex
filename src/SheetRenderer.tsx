import React, {Component, ReactNode} from 'react';
import RDS from 'react-datasheet';

import {
  ColumnHeader,
  HeaderContentRenderer,
  HeaderRenderer,
  SheetRendererProps,
} from './interfaces';
import DefaultHeaderRenderer from './DefaultHeaderRenderer';
import DefaultHeaderContentRenderer from './DefaultHeaderContentRenderer';

// import './react-datasheet-ex.css';


/**
 * A custom `sheetRenderer` for `react-datasheet` that renders a sheet
 * with column headers and an leading 'action" cell on each row.
 */
export default class SheetRenderer<
    T extends RDS.Cell<T, V>, V = string, H extends ColumnHeader = ColumnHeader
  > extends Component<SheetRendererProps<T, V, H>> {

  static displayName = 'SheetRenderer';

  static readonly defaultProps = {
    headers: [],
    headerRenderer: DefaultHeaderRenderer,
    headerContentRenderer: DefaultHeaderContentRenderer,
  };

  render(): ReactNode {
    const {
      className,
      headers,
      headerRenderer: HeaderRenderer,
      headerContentRenderer: HeaderContentRenderer,
      style,
    } = this.props;

    return (
      <table className={className} style={style}>
        <thead>
          <tr key='$headers'>
            <th className='cell read-only rdr-sheet-renderer__action-cell' key='$$actionCell'/>
            {headers.map((header: H, col: number) => {
              const hrProps = {col, header};
              return (
                <HeaderRenderer key={`th-${header.id}`} {...hrProps}>
                  <HeaderContentRenderer key={header.id} {...hrProps} />
                </HeaderRenderer>
              );
            })}
          </tr>
        </thead>
        <tbody>
        {this.props.children}
        </tbody>
      </table>
    );
  }
}
