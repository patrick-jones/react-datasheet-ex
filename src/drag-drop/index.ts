import DragDropHeaderProvider from './DragDropHeaderProvider';
import DragDropHeaderRenderer from './DragDropHeaderRenderer';
import DragDropRowProvider from './DragDropRowProvider';
import DragDropRowRenderer from './DragDropRowRenderer';

export {DragDropHeaderProvider, DragDropHeaderRenderer, DragDropRowProvider, DragDropRowRenderer};
