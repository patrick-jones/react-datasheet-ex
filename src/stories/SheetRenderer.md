# SheetRenderer

A custom `sheetRenderer` for `react-datasheet` that adds
column headers and a leading "action" cell on each row. 
It should be used with `RowRenderer` (from this package) 
or another a row renderer that adds an "action" cell 
at the beginning of each row.

*Note:* The columns headers and action cell are not considered part of 
the data grid's `data` array.
 

<!-- STORY -->

## Prop Types

### SheetRenderer

In addition to the 
[standard `sheetRenderer` properties](https://github.com/nadbm/react-datasheet#sheet-renderer)
provided by `react-datasheet`, you must supply these additional properties.

Name | Type | Default | Description
:--- | :--- | :--- | :---:
headers               | `ColumnHeader[]`        | [] | A list of `ColumnHeader` objects, labels for the columns in the grid
headerRenderer        | `HeaderRenderer`        | `DefaultHeaderRenderer` | Optional function or component to render the header container (e.g., a `th` element)
headerContentRenderer | `HeaderContentRenderer` | `DefaultHeaderContentRenderer` | Optional function or component to render the header contents. The default renders the header's `title` property. 
style?                | object                  | - | Optional style object applied to the data grid's main `table` element.

### interface ColumnHeader

The minimum properties for a column header. This can be extended.

Name | Type | Default | Description
:--- | :--- | :--- | :---:
id     | string  | -  | A unique (within the grid) identifier for the column
title  | string  | -  | The column title, displayed by default in the header
width? | number  | -  | Optional. The column's width in pixels


### type HeaderRenderer

A component or function that renders the header container 
(e.g., a `th`).

Name | Type | Default | Description
:--- | :--- | :--- | :---:
col     | number          | -  | The 0-based column index
header  | `ColumnHeader`  | -  | A `ColumnHeader`


### type HeaderContentRenderer

A component or function that renders the header content.

Name | Type | Default | Description
:--- | :--- | :--- | :---:
col     | number          | -  | The 0-based column index
header  | `ColumnHeader`  | -  | A `ColumnHeader`


### RowRenderer

This component differs from the default renderer in `react-datasheet` in that it
adds a blank cell at the beginning of each row. This extra cell is used as the
drag handle 
 
In addition to the
[standard `rowRenderer` properties](https://github.com/nadbm/react-datasheet#row-renderer)
supplied by `react-datasheet`, this component can accept a custom renderer for
the action cell's content.

Name | Type | Default | Description
:--- | :--- | :--- | :---:
actionContentRenderer? | `ActionContentRenderer` | - | Optional. A function or component that renders the content of the action cell; default is to leave it empty 


### type ActionContentRenderer

A component or function that renders the row action cell content. 

Name | Type | Default | Description
:--- | :--- | :--- | :---:
row | number | - | The 0-based row index

## Sample code

```js
import {SheetRenderer, RowRenderer} from 'react-datasheet-ex';

class Example extends Component {
  
  
  sheetRenderer = sheetRendererProps => (
    // add extra properties here
    <SheetRenderer {...sheetRendererProps} headers={this.props.headers} />
  );
  
  
  rowRenderer = rowRendererProps => (
    <RowRenderer {...rowRendererProps} />
  );
  
  
  render() {
    const {rows} = this.props;
    return (
      <ReactDatasheet
        data={rows}
        sheetRenderer={this.sheetRenderer}
        rowRenderer={this.rowRenderer}
        valueRenderer={cell => cell.value}
      />
    );
  }
  
}
```
