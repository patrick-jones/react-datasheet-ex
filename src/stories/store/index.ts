import {initialState} from './initialState';

export {initialState};
