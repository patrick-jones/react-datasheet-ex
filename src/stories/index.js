import React from 'react';
import {storiesOf} from '@storybook/react';
import {withState} from '@dump247/storybook-state';
import {doc} from 'storybook-readme';
import {withDocs} from './withDocs';

import DragDropProvidersStory from './DragDropProviders.story';
import DragDropRenderersStory from './DragDropRenderers.story';
import SheetRendererStory from './SheetRenderer.story';

import {initialState} from './store';

import '!style-loader!css-loader!react-datasheet/lib/react-datasheet.css';
import '../react-datasheet-ex.css';
import './stories.css';

import Readme from '../../README.md';
import SheetRendererDocs from './SheetRenderer.md';
import DragDropProvidersDocs from './DragDropProviders.md';
import DragDropRenderersDocs from './DragDropRenderers.md';

const stories = storiesOf('ReactDatasheet Extensions', module);

stories
  .add('Introduction', doc(Readme))
  .add('SheetRenderer - Basic',
    withDocs(SheetRendererDocs,
      withState(initialState)(SheetRendererStory)))
  .add('Drag and Drop Renderers',
    withDocs(DragDropRenderersDocs,
      withState(initialState)((DragDropRenderersStory))))
  .add('Drag and Drop Providers',
    withDocs(DragDropProvidersDocs,
      withState(initialState)((DragDropProvidersStory))))
;
