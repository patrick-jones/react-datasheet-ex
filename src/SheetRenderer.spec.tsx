import React from 'react';
import {shallow} from 'enzyme';

import SheetRenderer from './SheetRenderer';
import {DragDropHeaderRenderer} from './drag-drop';
import {initialState} from './stories/store';
import {ExampleCellType} from './stories/store/interfaces';


describe('SheetRenderer basic rendering', () => {
  const {rows, headers} = initialState;
  const data: ExampleCellType[][] = rows.map(r => r.data);

  it('Should render', () => {
    const wrapper = shallow(
      <SheetRenderer
        data={data}
        className='foobar'
        headers={headers}
      >
        <tr>
          <td colSpan={4}>shallow</td>
        </tr>
      </SheetRenderer>
    );
    // console.log(wrapper.debug());
    expect(wrapper.find('thead tr').length).toBe(1);
    expect(wrapper.find('th').length).toBe(1);
    expect(wrapper.find('DefaultHeaderRenderer').length).toBe(4);
    expect(wrapper.find('DefaultHeaderContentRenderer').length).toBe(4);
    expect(wrapper.find('tbody tr td').text()).toBe('shallow');
  });

  it('Matches the snapshot', () => {
    const wrapper = shallow(
      <SheetRenderer
        data={data}
        className='foobar'
        headers={headers}
      >
        <tr>
          <td colSpan={4}>shallow</td>
        </tr>
      </SheetRenderer>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('Correctly renders optional properties', () => {
    const wrapper = shallow(
      <SheetRenderer
        data={data}
        className='foobar'
        style={{width: '100%'}}
        headers={headers}
        headerRenderer={hrProps => (
          <DragDropHeaderRenderer {...hrProps} onHeaderDropped={(...args) => {}} />
        )}
        headerContentRenderer={hcrProps => (<span>{hcrProps.header.id}</span>)}
      >
        <tr>
          <td colSpan={4}>shallow</td>
        </tr>
      </SheetRenderer>
    );
    // console.log(wrapper.debug());
    expect(wrapper.find('thead tr').length).toBe(1);
    expect(wrapper.find('th').length).toBe(1);
    expect(wrapper.find('headerRenderer').length).toBe(4);
    expect(wrapper.find('headerContentRenderer').length).toBe(4);
    expect(wrapper.find('tbody tr td').text()).toBe('shallow');
    expect(wrapper.prop('style')).toEqual({width: '100%'});
  });
});
