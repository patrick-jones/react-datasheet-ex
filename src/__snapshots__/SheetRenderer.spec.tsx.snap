// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SheetRenderer basic rendering Matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SheetRenderer
    className="foobar"
    data={
      Array [
        Array [
          Object {
            "value": 1,
          },
          Object {
            "value": 2,
          },
          Object {
            "value": 3,
          },
          Object {
            "value": 4,
          },
        ],
        Array [
          Object {
            "value": 5,
          },
          Object {
            "value": 6,
          },
          Object {
            "value": 7,
          },
          Object {
            "value": 8,
          },
        ],
        Array [
          Object {
            "value": 9,
          },
          Object {
            "value": 10,
          },
          Object {
            "value": 11,
          },
          Object {
            "value": 12,
          },
        ],
      ]
    }
    headerContentRenderer={[Function]}
    headerRenderer={[Function]}
    headers={
      Array [
        Object {
          "id": "A",
          "title": "A",
        },
        Object {
          "id": "B",
          "title": "B",
        },
        Object {
          "id": "C",
          "title": "C",
        },
        Object {
          "id": "D",
          "title": "D",
        },
      ]
    }
  >
    <tr>
      <td
        colSpan={4}
      >
        shallow
      </td>
    </tr>
  </SheetRenderer>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <thead>
          <tr>
            <th
              className="cell read-only rdr-sheet-renderer__action-cell"
            />
            <DefaultHeaderRenderer
              col={0}
              header={
                Object {
                  "id": "A",
                  "title": "A",
                }
              }
            >
              <DefaultHeaderContentRenderer
                col={0}
                header={
                  Object {
                    "id": "A",
                    "title": "A",
                  }
                }
              />
            </DefaultHeaderRenderer>
            <DefaultHeaderRenderer
              col={1}
              header={
                Object {
                  "id": "B",
                  "title": "B",
                }
              }
            >
              <DefaultHeaderContentRenderer
                col={1}
                header={
                  Object {
                    "id": "B",
                    "title": "B",
                  }
                }
              />
            </DefaultHeaderRenderer>
            <DefaultHeaderRenderer
              col={2}
              header={
                Object {
                  "id": "C",
                  "title": "C",
                }
              }
            >
              <DefaultHeaderContentRenderer
                col={2}
                header={
                  Object {
                    "id": "C",
                    "title": "C",
                  }
                }
              />
            </DefaultHeaderRenderer>
            <DefaultHeaderRenderer
              col={3}
              header={
                Object {
                  "id": "D",
                  "title": "D",
                }
              }
            >
              <DefaultHeaderContentRenderer
                col={3}
                header={
                  Object {
                    "id": "D",
                    "title": "D",
                  }
                }
              />
            </DefaultHeaderRenderer>
          </tr>
        </thead>,
        <tbody>
          <tr>
            <td
              colSpan={4}
            >
              shallow
            </td>
          </tr>
        </tbody>,
      ],
      "className": "foobar",
      "style": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <tr>
            <th
              className="cell read-only rdr-sheet-renderer__action-cell"
            />
            <DefaultHeaderRenderer
              col={0}
              header={
                Object {
                  "id": "A",
                  "title": "A",
                }
              }
            >
              <DefaultHeaderContentRenderer
                col={0}
                header={
                  Object {
                    "id": "A",
                    "title": "A",
                  }
                }
              />
            </DefaultHeaderRenderer>
            <DefaultHeaderRenderer
              col={1}
              header={
                Object {
                  "id": "B",
                  "title": "B",
                }
              }
            >
              <DefaultHeaderContentRenderer
                col={1}
                header={
                  Object {
                    "id": "B",
                    "title": "B",
                  }
                }
              />
            </DefaultHeaderRenderer>
            <DefaultHeaderRenderer
              col={2}
              header={
                Object {
                  "id": "C",
                  "title": "C",
                }
              }
            >
              <DefaultHeaderContentRenderer
                col={2}
                header={
                  Object {
                    "id": "C",
                    "title": "C",
                  }
                }
              />
            </DefaultHeaderRenderer>
            <DefaultHeaderRenderer
              col={3}
              header={
                Object {
                  "id": "D",
                  "title": "D",
                }
              }
            >
              <DefaultHeaderContentRenderer
                col={3}
                header={
                  Object {
                    "id": "D",
                    "title": "D",
                  }
                }
              />
            </DefaultHeaderRenderer>
          </tr>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": "$headers",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <th
                className="cell read-only rdr-sheet-renderer__action-cell"
              />,
              Array [
                <DefaultHeaderRenderer
                  col={0}
                  header={
                    Object {
                      "id": "A",
                      "title": "A",
                    }
                  }
                >
                  <DefaultHeaderContentRenderer
                    col={0}
                    header={
                      Object {
                        "id": "A",
                        "title": "A",
                      }
                    }
                  />
                </DefaultHeaderRenderer>,
                <DefaultHeaderRenderer
                  col={1}
                  header={
                    Object {
                      "id": "B",
                      "title": "B",
                    }
                  }
                >
                  <DefaultHeaderContentRenderer
                    col={1}
                    header={
                      Object {
                        "id": "B",
                        "title": "B",
                      }
                    }
                  />
                </DefaultHeaderRenderer>,
                <DefaultHeaderRenderer
                  col={2}
                  header={
                    Object {
                      "id": "C",
                      "title": "C",
                    }
                  }
                >
                  <DefaultHeaderContentRenderer
                    col={2}
                    header={
                      Object {
                        "id": "C",
                        "title": "C",
                      }
                    }
                  />
                </DefaultHeaderRenderer>,
                <DefaultHeaderRenderer
                  col={3}
                  header={
                    Object {
                      "id": "D",
                      "title": "D",
                    }
                  }
                >
                  <DefaultHeaderContentRenderer
                    col={3}
                    header={
                      Object {
                        "id": "D",
                        "title": "D",
                      }
                    }
                  />
                </DefaultHeaderRenderer>,
              ],
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "$$actionCell",
              "nodeType": "host",
              "props": Object {
                "className": "cell read-only rdr-sheet-renderer__action-cell",
              },
              "ref": null,
              "rendered": null,
              "type": "th",
            },
            Object {
              "instance": null,
              "key": "th-A",
              "nodeType": "class",
              "props": Object {
                "children": <DefaultHeaderContentRenderer
                  col={0}
                  header={
                    Object {
                      "id": "A",
                      "title": "A",
                    }
                  }
                />,
                "col": 0,
                "header": Object {
                  "id": "A",
                  "title": "A",
                },
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": "A",
                "nodeType": "class",
                "props": Object {
                  "col": 0,
                  "header": Object {
                    "id": "A",
                    "title": "A",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "th-B",
              "nodeType": "class",
              "props": Object {
                "children": <DefaultHeaderContentRenderer
                  col={1}
                  header={
                    Object {
                      "id": "B",
                      "title": "B",
                    }
                  }
                />,
                "col": 1,
                "header": Object {
                  "id": "B",
                  "title": "B",
                },
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": "B",
                "nodeType": "class",
                "props": Object {
                  "col": 1,
                  "header": Object {
                    "id": "B",
                    "title": "B",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "th-C",
              "nodeType": "class",
              "props": Object {
                "children": <DefaultHeaderContentRenderer
                  col={2}
                  header={
                    Object {
                      "id": "C",
                      "title": "C",
                    }
                  }
                />,
                "col": 2,
                "header": Object {
                  "id": "C",
                  "title": "C",
                },
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": "C",
                "nodeType": "class",
                "props": Object {
                  "col": 2,
                  "header": Object {
                    "id": "C",
                    "title": "C",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "th-D",
              "nodeType": "class",
              "props": Object {
                "children": <DefaultHeaderContentRenderer
                  col={3}
                  header={
                    Object {
                      "id": "D",
                      "title": "D",
                    }
                  }
                />,
                "col": 3,
                "header": Object {
                  "id": "D",
                  "title": "D",
                },
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": "D",
                "nodeType": "class",
                "props": Object {
                  "col": 3,
                  "header": Object {
                    "id": "D",
                    "title": "D",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": "tr",
        },
        "type": "thead",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <tr>
            <td
              colSpan={4}
            >
              shallow
            </td>
          </tr>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <td
              colSpan={4}
            >
              shallow
            </td>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "shallow",
              "colSpan": 4,
            },
            "ref": null,
            "rendered": "shallow",
            "type": "td",
          },
          "type": "tr",
        },
        "type": "tbody",
      },
    ],
    "type": "table",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <thead>
            <tr>
              <th
                className="cell read-only rdr-sheet-renderer__action-cell"
              />
              <DefaultHeaderRenderer
                col={0}
                header={
                  Object {
                    "id": "A",
                    "title": "A",
                  }
                }
              >
                <DefaultHeaderContentRenderer
                  col={0}
                  header={
                    Object {
                      "id": "A",
                      "title": "A",
                    }
                  }
                />
              </DefaultHeaderRenderer>
              <DefaultHeaderRenderer
                col={1}
                header={
                  Object {
                    "id": "B",
                    "title": "B",
                  }
                }
              >
                <DefaultHeaderContentRenderer
                  col={1}
                  header={
                    Object {
                      "id": "B",
                      "title": "B",
                    }
                  }
                />
              </DefaultHeaderRenderer>
              <DefaultHeaderRenderer
                col={2}
                header={
                  Object {
                    "id": "C",
                    "title": "C",
                  }
                }
              >
                <DefaultHeaderContentRenderer
                  col={2}
                  header={
                    Object {
                      "id": "C",
                      "title": "C",
                    }
                  }
                />
              </DefaultHeaderRenderer>
              <DefaultHeaderRenderer
                col={3}
                header={
                  Object {
                    "id": "D",
                    "title": "D",
                  }
                }
              >
                <DefaultHeaderContentRenderer
                  col={3}
                  header={
                    Object {
                      "id": "D",
                      "title": "D",
                    }
                  }
                />
              </DefaultHeaderRenderer>
            </tr>
          </thead>,
          <tbody>
            <tr>
              <td
                colSpan={4}
              >
                shallow
              </td>
            </tr>
          </tbody>,
        ],
        "className": "foobar",
        "style": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <tr>
              <th
                className="cell read-only rdr-sheet-renderer__action-cell"
              />
              <DefaultHeaderRenderer
                col={0}
                header={
                  Object {
                    "id": "A",
                    "title": "A",
                  }
                }
              >
                <DefaultHeaderContentRenderer
                  col={0}
                  header={
                    Object {
                      "id": "A",
                      "title": "A",
                    }
                  }
                />
              </DefaultHeaderRenderer>
              <DefaultHeaderRenderer
                col={1}
                header={
                  Object {
                    "id": "B",
                    "title": "B",
                  }
                }
              >
                <DefaultHeaderContentRenderer
                  col={1}
                  header={
                    Object {
                      "id": "B",
                      "title": "B",
                    }
                  }
                />
              </DefaultHeaderRenderer>
              <DefaultHeaderRenderer
                col={2}
                header={
                  Object {
                    "id": "C",
                    "title": "C",
                  }
                }
              >
                <DefaultHeaderContentRenderer
                  col={2}
                  header={
                    Object {
                      "id": "C",
                      "title": "C",
                    }
                  }
                />
              </DefaultHeaderRenderer>
              <DefaultHeaderRenderer
                col={3}
                header={
                  Object {
                    "id": "D",
                    "title": "D",
                  }
                }
              >
                <DefaultHeaderContentRenderer
                  col={3}
                  header={
                    Object {
                      "id": "D",
                      "title": "D",
                    }
                  }
                />
              </DefaultHeaderRenderer>
            </tr>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": "$headers",
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <th
                  className="cell read-only rdr-sheet-renderer__action-cell"
                />,
                Array [
                  <DefaultHeaderRenderer
                    col={0}
                    header={
                      Object {
                        "id": "A",
                        "title": "A",
                      }
                    }
                  >
                    <DefaultHeaderContentRenderer
                      col={0}
                      header={
                        Object {
                          "id": "A",
                          "title": "A",
                        }
                      }
                    />
                  </DefaultHeaderRenderer>,
                  <DefaultHeaderRenderer
                    col={1}
                    header={
                      Object {
                        "id": "B",
                        "title": "B",
                      }
                    }
                  >
                    <DefaultHeaderContentRenderer
                      col={1}
                      header={
                        Object {
                          "id": "B",
                          "title": "B",
                        }
                      }
                    />
                  </DefaultHeaderRenderer>,
                  <DefaultHeaderRenderer
                    col={2}
                    header={
                      Object {
                        "id": "C",
                        "title": "C",
                      }
                    }
                  >
                    <DefaultHeaderContentRenderer
                      col={2}
                      header={
                        Object {
                          "id": "C",
                          "title": "C",
                        }
                      }
                    />
                  </DefaultHeaderRenderer>,
                  <DefaultHeaderRenderer
                    col={3}
                    header={
                      Object {
                        "id": "D",
                        "title": "D",
                      }
                    }
                  >
                    <DefaultHeaderContentRenderer
                      col={3}
                      header={
                        Object {
                          "id": "D",
                          "title": "D",
                        }
                      }
                    />
                  </DefaultHeaderRenderer>,
                ],
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "$$actionCell",
                "nodeType": "host",
                "props": Object {
                  "className": "cell read-only rdr-sheet-renderer__action-cell",
                },
                "ref": null,
                "rendered": null,
                "type": "th",
              },
              Object {
                "instance": null,
                "key": "th-A",
                "nodeType": "class",
                "props": Object {
                  "children": <DefaultHeaderContentRenderer
                    col={0}
                    header={
                      Object {
                        "id": "A",
                        "title": "A",
                      }
                    }
                  />,
                  "col": 0,
                  "header": Object {
                    "id": "A",
                    "title": "A",
                  },
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": "A",
                  "nodeType": "class",
                  "props": Object {
                    "col": 0,
                    "header": Object {
                      "id": "A",
                      "title": "A",
                    },
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "th-B",
                "nodeType": "class",
                "props": Object {
                  "children": <DefaultHeaderContentRenderer
                    col={1}
                    header={
                      Object {
                        "id": "B",
                        "title": "B",
                      }
                    }
                  />,
                  "col": 1,
                  "header": Object {
                    "id": "B",
                    "title": "B",
                  },
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": "B",
                  "nodeType": "class",
                  "props": Object {
                    "col": 1,
                    "header": Object {
                      "id": "B",
                      "title": "B",
                    },
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "th-C",
                "nodeType": "class",
                "props": Object {
                  "children": <DefaultHeaderContentRenderer
                    col={2}
                    header={
                      Object {
                        "id": "C",
                        "title": "C",
                      }
                    }
                  />,
                  "col": 2,
                  "header": Object {
                    "id": "C",
                    "title": "C",
                  },
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": "C",
                  "nodeType": "class",
                  "props": Object {
                    "col": 2,
                    "header": Object {
                      "id": "C",
                      "title": "C",
                    },
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "th-D",
                "nodeType": "class",
                "props": Object {
                  "children": <DefaultHeaderContentRenderer
                    col={3}
                    header={
                      Object {
                        "id": "D",
                        "title": "D",
                      }
                    }
                  />,
                  "col": 3,
                  "header": Object {
                    "id": "D",
                    "title": "D",
                  },
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": "D",
                  "nodeType": "class",
                  "props": Object {
                    "col": 3,
                    "header": Object {
                      "id": "D",
                      "title": "D",
                    },
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
            ],
            "type": "tr",
          },
          "type": "thead",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <tr>
              <td
                colSpan={4}
              >
                shallow
              </td>
            </tr>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <td
                colSpan={4}
              >
                shallow
              </td>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "shallow",
                "colSpan": 4,
              },
              "ref": null,
              "rendered": "shallow",
              "type": "td",
            },
            "type": "tr",
          },
          "type": "tbody",
        },
      ],
      "type": "table",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
